### DISCUSSION TITLE
Pulling all mockups from a folder

### DISCUSSION THREAD

**Discussion by zgdesign** [joonaspaakko/Batch-Mockup-Smart-Object-Replacement-photoshop-script#21] [Q&A] [answered]
Hello,
First of all I'd like to thank you for this amazing script. It's a massive timesaver, and very customizable.

I have quite a few mockups and am trying to figure out a way to pull all the mockups from a folder, instead of creating a new mockup object for each one. I tried using an array for the mockupPath, but that doesn't seem to work.

It would be ideal if there was a way to point to a folder for the script to look for all the mockup files from within a folder, sort of like the input folder. 

When you have 60+ mockups, duplicating the objects gets pretty messy and time consuming.

Thanks!

**Comment by joonaspaakko** [comment, marked as answer]
Yea, this isn't an existing feature, but the settings script can be extended with javascript. I made you an example that should do it, or at least get you pretty far. Though I should say I didn't really test it thoroughly, and keep in mind it requires each mockup to have the same exact settings. Most importantly all mockup files need to have the same name for the smart object target layer name.

> **Note:** I made a similar, but more manual example at the bottom of [this answer](https://github.com/joonaspaakko/Batch-Mockup-Smart-Object-Replacement-photoshop-script/discussions/18#discussioncomment-4876787). The big difference being that in it you can customize each mockup settings as much as you need or as little as you want.

You can test it with the [standalone example](https://github.com/joonaspaakko/Batch-Mockup-Smart-Object-Replacement.js-Photoshop-script-standalone-example) by replacing the settings file's contents with the following code:
- `output.folders` is set to `true` so that's it easier to see the script processed all mockup files
- Also you should duplicate the `assets/example-mockup.psd` file a couple of times so you can see the script automatically picking up all PSD files in the `assets` folder.

```js
#include "script/Batch Mockup Smart Object Replacement.jsx" 

var output = {
  path: './_output files', // Outputs next to the mockup psd
  format: 'jpg', // 'jpg', 'png', 'tif', 'psd', 'pdf'
  filename: '@mockup - @input', 
  folders: true,
};

// Later on after it's been filled with mockup settings this array is fed to the mockups() method
// that starts the mockup batch process...
var mockupsArray = [];

// This method returns the settings object for a single mockup and replaces the "mockupPath" with
// the input "path".
var addMockup = function( path ) {
  
  var mockup = {
    output: output, 
    mockupPath: path, 
    smartObjects: [

      {
        target: '@screen',
        input: [
          './_input files'
        ], 
        inputNested: true,
        align: 'center center',
        resize: 'fill', 
      },
      
    ]
  };
  
  return mockup;
  
};

// Get path to current file
var pathToThisFile = scriptFolder();
// Append relative path to the mockups...
var mockupsFolder = new Folder( pathToThisFile + 'assets' );
// Get mockup files in the specified folder (psd or psb)
var mockupFiles = getFiles( mockupsFolder, { inputFormats: "psd|psb", inputNested: false });

// Goes through every found mockup file...
each( mockupFiles, function( mockupFile ) {
  
  // Get text string of the full (absolute) path to the file
  var pathToMockup = mockupFile.fullName;
  // Create new mockup settings object using the path...
  var newMockupSettings = addMockup( pathToMockup );
  // Add new mockup settings to the array of mockups
  mockupsArray.push( newMockupSettings );
  
});

// scriptFolder(), getFiles, and each() are methods that come from the #included engine script

// This mockups() method is the method that you would normally fill up with objects manuall.
// With this setup you can still add mockup settings manually by pushing to the "mockupsArray"
mockups( mockupsArray );
```

**Reply by zgdesign** [reply]
This is exactly what I needed, thank you so much for taking the time to write the extra code and for such a quick reply!

**Reply by joonaspaakko** [reply]
I added this v.1.9. (fc546ba). All you need is point `mockupPath` to a folder rather than a PSD file. Additionally there's a new mockup option `mockupNested` that works exactly the same as `inputNested`, making it pull all mockup files in the `mockupPath` folder.

**Reply by zgdesign** [reply]
I was hoping you could help me out with an issue. I could've sworn this was working before, but now I just can't get the script to pull all the mockups from a folder. It works just fine if I point it directly to a mockup file, but if I leave it as just a folder path and set the mockupNested set to true, all I get is a "Batch process done!" alert, but nothing actually happens. If I change it to the folder path and mockup file name it works, but just on that single mockup.

        mockupPath: '$/edited/mockups', // this doesn't work
        mockupPath: '$/edited/mockups/01.tif', // this works but just on a single mockup like in previous version

#include "Batch Mockup Smart Object Replacement.jsx"

var output = {
    path: '$/BATCH OUTPUT', // Outputs next to the mockup psd
    format: 'jpg', // 'jpg', 'png', 'tif', 'psd', 'pdf'
    filename: '@mockup-@input',
};

mockups([

    // Mockup #1
    {
        output: {
            path: '$/BATCH OUTPUT',
            format: 'jpg', // 'jpg', 'png', 'tif', 'psd', 'pdf'
            folders: false, // Files will be grouped by folders inside the output folder
            filename: '@mockup-@input', 
        },
        //mockupNested: true,
        mockupPath: '$/edited/mockups', // this doesn't work
        mockupNested: true,
        input: ['../../../PRINT PRODUCTION/LASER FILES/ornaments/led ornaments',],
        
    smartObjects: [

        // Smart object #1
        {
            target: 'art',
            input: ['../../../PRINT PRODUCTION/LASER FILES/ornaments/led ornaments',],
        
        //inputNested: true,
        align: 'center center',
        resize: 'fit', 
        },

    ]
  },

    // You could have more mockup files...
    // {},

]);

**Reply by joonaspaakko** [reply]
I was thinking that maybe there's a bug. I let this simmer overnight, and I woke up with a realization that you said a direct path to `.tif` works... But when you use a path to a folder, it's hardcoded to find `psd` or `psb` files. The tiff file format is pretty much nonexistent in the premade mockups space (though honestly so is `psb`), so it didn't even occur to me to include `tif`. 

Rather than add that as one of the defaults, I think I should add a `mockupFormats` option, just like the `inputFormats`, so you could easily pick what file types you want to target with the `mockupPath`. 

https://github.com/joonaspaakko/Batch-Mockup-Smart-Object-Replacement-photoshop-script/blob/fc546baacf6ea29cab95eb3b97cb869ae2baeccd/script/Batch%20Mockup%20Smart%20Object%20Replacement.jsx#L164

For now you could change this ↑ line to: 
```js
 var mockupFiles = getFiles( mockupsFolder, { inputFormats: "psd|psb|tiff?", inputNested: !!mockup.mockupNested }); 
```
> Or I guess just `inputFormats: "tiff?"` if you never use PSD or PSB. The question mark makes it so the previous character is optional, so it matches `tif` and `tiff` files. For reference, the default `inputFormats` for input files are `tiff?|gif|jpe?g|bmp|eps|svg|png|ai|psd|pdf'`. Interestingly I excluded `psb` here, I guess I just forgot.

**Comment by zgdesign** [comment]
It never occurred to me that it could be because of the file type. I thought perhaps I had one of the paths mistyped.
Once again you've come to the rescue, it works perfectly now!
The only reason I use TIFs vs PSDs is because of the compression options. When you have thousands of hi-res files it adds up pretty quickly.
Thanks again for your help!
